From a7f437811df15dbb0395c2a228a97f86ee329b7f Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 23 May 2022 20:33:59 +0300
Subject: [PATCH] [SW-0] gaudi3: update checks for entering recovery mode

eROM recovery will be executed based on three efuse value
1. RECOVERY LOCATION bit set to 0
2. OOB protocol bit set to 1
3. I3C EROM bit set to 1

This patch update checks as mentioned above.

Change-Id: Iea16fd890848755010703b9bc26b1f674f254976
---

diff --git a/arch/gaudi3/arc/dev_efuse.h b/arch/gaudi3/arc/dev_efuse.h
index f96dbf0..bd029c0 100644
--- a/arch/gaudi3/arc/dev_efuse.h
+++ b/arch/gaudi3/arc/dev_efuse.h
@@ -21,9 +21,12 @@
 #define EFUSE_SYSCFG_SECURITY_ENABLE_MASK		BIT(15)
 #define EFUSE_SYSCFG_RR_MASK				BIT(16)
 #define EFUSE_SYSCFG_ENABLE_TPM_MASK			BIT(17)
+ /* TODO: confirm bit for OOB protocol */
+#define EFUSE_SYSCFG_OOB_PROTOCOL			BIT(19)
 #define EFUSE_SYSCFG_BACKUP_DISABLE			BIT(22)
 #define EFUSE_SYSCFG_SKU_MASK				BIT(26)
 #define EFUSE_SYSCFG_PCIE_CONTROL_MASK			BIT(29)
+#define EFUSE_SYSCFG_I3C_EROM_ENABLE_MASK		BIT(31)
 #define EFUSE_SYSCFG2_D2D_DIS_MASK			BIT(31)
 
 /* Following bitmaps are used in common code, hence maintained for compilation */
diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index 0436a3b..a4ff4e1 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -301,14 +301,22 @@
 		arcboot_rom_notify_slave();
 
 	/*
-	 * efuse[7]:
+	 * efuse[EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_MASK]:
 	 * 0: eROM
 	 * 1: ROM
+	 * efuse[EFUSE_SYSCFG_OOB_PROTOCOL]:
+	 * 0: NVMe (Legacy mode) -> ROM
+	 * 1: PLDM -> eROM
+	 * efuse[EFUSE_SYSCFG_I3C_EROM_ENABLE_MASK]:
+	 * 0: No I3C -> ROM
+	 * 1: I3C -> eROM
 	 */
-	if (arcboot_rom_global.config_fuse & EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_MASK)
-		execute_rom_recovery();
-	else
+	if ((!(arcboot_rom_global.config_fuse & EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_MASK)) ||
+		(arcboot_rom_global.config_fuse & EFUSE_SYSCFG_OOB_PROTOCOL) ||
+		(arcboot_rom_global.config_fuse & EFUSE_SYSCFG_I3C_EROM_ENABLE_MASK))
 		execute_erom_recovery();
+	else
+		execute_rom_recovery();
 
 	/*
 	 * ROM Execution is expected to reach here only in case EROM image
